backend/de.metas.async/src/main/java/de/metas/async/processor/impl/planner/QueueProcessorPlanner.java

```diff
@@ -0,0 +1,360 @@
+
+import java.util.List;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.stream.Collectors;
+
+public abstract class QueueProcessorPlanner implements Runnable
+{
+	private static final Logger logger = LogManager.getLogger(QueueProcessorPlanner.class);
+
+	private final static String SYSCONFIG_POLLINTERVAL = "de.metas.async.PollIntervallMillis";
+
+	private final ILockManager lockManager = Services.get(ILockManager.class);
+	private final IQueryBL queryBL = Services.get(IQueryBL.class);
+
+	private final ReentrantLock mainLock = new ReentrantLock();
+	protected final ConcurrentHashMap<QueueProcessorId, IQueueProcessor> queueProcessors;
+	protected final AtomicBoolean isRunning;
+
+	public QueueProcessorPlanner()
+	{
+		this.queueProcessors = new ConcurrentHashMap<>();
+		this.isRunning = new AtomicBoolean(false);
+	}
+
+	@Override
+	public void run()
+	{
+		if (!isRunning.get())
+		{
+			throw new AdempiereException("Planner is not running! Hint: QueueProcessorPlanner should be started via start() method!");
+		}
+
+		while (isRunning.get())
+		{
+			try
+			{
+				final boolean successfulRun = run0();
+
+				if (!successfulRun && stopOnFailedRun())
+				{
+					logger.warn("*** QueueProcessorPlanner.run() -> last run was not successful & planner = " + this.getClass().getName() + " has stopOnFailedRun policy; Shutting down...");
+					shutdown();
+				}
+				else
+				{
+					// note: we always get the new service, because things might have changed since this method started
+					final int pollIntervalMs = Services.get(ISysConfigBL.class).getIntValue(SYSCONFIG_POLLINTERVAL, 1000);
+
+					Thread.sleep(pollIntervalMs);
+				}
+			}
+			catch (final InterruptedException e)
+			{
+				logger.error("Got interrupt request.", e);
+				return;
+			}
+			catch (final Throwable throwable)
+			{
+				logger.warn("*** QueueProcessorPlanner.run() caught an unexpected error!", throwable);
+
+				if (stopOnFailedRun())
+				{
+					logger.error("*** QueueProcessorPlanner.run() caught an unexpected error!"
+										 + " Planner = " + this.getClass().getName() + " has stopOnFailedRun policy; Shutting down...", throwable);
+					shutdown();
+
+					throw AdempiereException.wrapIfNeeded(throwable);
+				}
+			}
+		}
+	}
+
+	@NonNull
+	public List<IQueueProcessor> getRegisteredQueueProcessors()
+	{
+		return ImmutableList.copyOf(queueProcessors.values());
+	}
+
+	public void addQueueProcessor(@NonNull final IQueueProcessor queueProcessor)
+	{
+		validateQueueProcessorAssignment(queueProcessor);
+
+		queueProcessors.put(queueProcessor.getQueueProcessorId(), queueProcessor);
+	}
+
+	public boolean removeQueueProcessor(@NonNull final QueueProcessorId queueProcessorId)
+	{
+		final Optional<IQueueProcessor> processorOpt = getQueueProcessor(queueProcessorId);
```

Comment:
imho i would avoid having optional variables.

this code could be written as
```java
getQueueProcessor(queueProcessorId).ifPresent(processor -> {
  processor.shutdownExecutor();
  queueProcessors.remove(processor.getQueueProcessorId());
});
```

(maybe) you could also extract the 
```java
  processor.shutdownExecutor();
  queueProcessors.remove(processor.getQueueProcessorId());
```
to a separate method, something like 'shutdownAndRemove(processor)`.